pkg/webhook/v1alpha2/applicationconfiguration/mutating_handler.go

```diff
@@ -0,0 +1,174 @@
+
+	err := h.Decoder.Decode(req, obj)
+	if err != nil {
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+	// mutate the object
+	if err := h.Mutate(obj); err != nil {
+		mutatelog.Error(err, "failed to mutate the applicationConfiguration", "name", obj.Name)
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+	mutatelog.Info("Print the mutated obj", "obj name", obj.Name, "mutated obj", spew.Sdump(obj.Spec))
+
+	marshalled, err := json.Marshal(obj)
+	if err != nil {
+		return admission.Errored(http.StatusInternalServerError, err)
+	}
+
+	resp := admission.PatchResponseFromRaw(req.AdmissionRequest.Object.Raw, marshalled)
+	if len(resp.Patches) > 0 {
+		mutatelog.Info("admit ApplicationConfiguration",
+			"namespace", obj.Namespace, "name", obj.Name, "patches", util.JSONMarshal(resp.Patches))
+	}
+	return resp
+}
+
+// Mutate sets all the default value for the Component
+func (h *MutatingHandler) Mutate(obj *v1alpha2.ApplicationConfiguration) error {
+	mutatelog.Info("mutate", "name", obj.Name)
+
+	for compIdx, comp := range obj.Spec.Components {
+		var updated bool
+		for idx, tr := range comp.Traits {
+			var content map[string]interface{}
+			if err := json.Unmarshal(tr.Trait.Raw, &content); err != nil {
+				return err
+			}
+			rawByte, mutated, err := h.mutateTrait(content, comp.ComponentName, obj.GetNamespace())
+			if err != nil {
+				return err
+			}
+			if !mutated {
+				continue
+			}
+			tr.Trait.Raw = rawByte
+			comp.Traits[idx] = tr
+			updated = true
+		}
+		if updated {
+			obj.Spec.Components[compIdx] = comp
+		}
+	}
+
+	return nil
+}
+
+func (h *MutatingHandler) mutateTrait(content map[string]interface{}, compName, namespace string) ([]byte, bool, error) {
+	if content[TraitTypeField] == nil {
+		return nil, false, nil
+	}
+	traitType, ok := content[TraitTypeField].(string)
+	if !ok {
+		return nil, false, fmt.Errorf("name of trait should be string instead of %s", reflect.TypeOf(content[TraitTypeField]))
+	}
+	mutatelog.Info("the trait refers to traitDefinition by name", "compName", compName, "trait name", traitType)
+	// Fetch the corresponding traitDefinition CR, the traitDefinition crd is cluster scoped
+	traitDefinition := &v1alpha2.TraitDefinition{}
+	if err := h.Client.Get(context.TODO(), types.NamespacedName{Name: traitType}, traitDefinition); err != nil {
+		return nil, false, err
+	}
+	// fetch the CRDs definition
+	customResourceDefinition := &crdv1.CustomResourceDefinition{}
+	if err := h.Client.Get(context.TODO(), types.NamespacedName{Name: traitDefinition.Spec.Reference.Name}, customResourceDefinition); err != nil {
+		return nil, false, err
+	}
+	// reconstruct the trait CR
+	delete(content, TraitTypeField)
+
+	if content[TraitSpecField] != nil {
+		content["spec"] = content[TraitSpecField]
+		delete(content, TraitSpecField)
+	}
+
+	trait := unstructured.Unstructured{
+		Object: content,
+	}
+	// find out the GVK from the CRD definition and set
+	apiVersion := metav1.GroupVersion{
```

Comment:
I am OK for short term fix by simply replacing name/properties with GVK. In long term it's better to fix this in AppConfig controller so `$ kubectl get appconfig` will also return name/properties format.

Reply:
Yes, this case is different from the `component` mutate hook as that was completely inside the component. I preferred to handle the explicit field in the controller instead. It's actually simpler to test.

Reply:
`component` mutate hook and this PR are actually the same, I hope we can use a separate PR to move them inside the controller together **if we really want to do that**.